File: scala/Solver/Tactics.scala
// ==========================================
// Tactics.scala
// タクティクスの実装（高度な書き換え・インスタンス化対応）
// ==========================================

package romanesco.Solver.core

import romanesco.Solver.core.Expr._
import romanesco.Solver.core.LogicSymbols._
import romanesco.Utils.Debug.logger

object Tactics {
  import Unifier._

  /** 含意(→)や全称量化(∀)の導入。
    */
  def intro(state: ProofState, name: Option[String] = None): TacticResult = {
    state.currentGoal match {
      case Some(Goal(ctx, lin, App(Sym(op), List(a, b))))
          if op == Implies || op == ImpliesAlt1 || op == ImpliesAlt2 =>
        val hypName = name.getOrElse(s"h${ctx.size}")
        val newGoal = Goal((hypName, a) :: ctx, lin, b)
        Right(state.copy(goals = newGoal :: state.goals.tail))

      // 線形含意(⊸)
      case Some(Goal(ctx, lin, App(Sym(LImplies), List(a, b)))) =>
        val hypName = name.getOrElse(s"lin${lin.size}")
        val newGoal = Goal(ctx, (hypName, a) :: lin, b)
        Right(state.copy(goals = newGoal :: state.goals.tail))

      case Some(Goal(ctx, lin, App(Sym(Forall), List(Var(v), body)))) =>
        val hypName = name.getOrElse(s"${v}_t${state.goals.size}")
        val instantiated = Prover.substVar(body, v, Var(hypName))
        val newGoal = Goal(ctx, lin, instantiated)
        Right(state.copy(goals = newGoal :: state.goals.tail))

      case _ =>
        Left("intro: Goal is not an implication or universal quantification")
    }
  }

  /** 連言(∧)の分解。
    */
  def split(state: ProofState): TacticResult = {
    state.currentGoal match {
      case Some(Goal(ctx, lin, App(Sym(op), List(a, b))))
          if op == And || op == Product =>
        val g1 = Goal(ctx, lin, a)
        val g2 = Goal(ctx, lin, b)
        Right(state.copy(goals = g1 :: g2 :: state.goals.tail))
      case _ => Left("split: Goal is not a conjunction")
    }
  }

  /** 仮定の分解。
    */
  def destruct(state: ProofState, hypName: String): TacticResult = {
    state.currentGoal match {
      case Some(goal @ Goal(ctx, lin, target)) =>
        // 持続的文脈の検索
        ctx.zipWithIndex.find(_._1._1 == hypName) match {
          case Some(((name, hypExpr), idx)) =>
            hypExpr match {
              case App(Sym(op), List(a, b)) if op == And || op == Product =>
                val newCtx =
                  ctx.patch(idx, List((s"$name.1", a), (s"$name.2", b)), 1)
                Right(
                  state.copy(goals =
                    goal.copy(context = newCtx) :: state.goals.tail
                  )
                )

              case App(Sym(op), List(a, b)) if op == Or || op == Coproduct =>
                val g1 =
                  Goal(ctx.patch(idx, List((s"$name.left", a)), 1), lin, target)
                val g2 =
                  Goal(ctx.patch(idx, List((s"$name.right", b)), 1), lin, target)
                Right(state.copy(goals = g1 :: g2 :: state.goals.tail))

              case _ =>
                Left(s"destruct: Hypothesis '$hypName' cannot be destructed")
            }
          case None => 
            // 線形文脈の検索
            lin.zipWithIndex.find(_._1._1 == hypName) match {
              case Some(((name, hypExpr), idx)) =>
                hypExpr match {
                  case App(Sym(Tensor), List(a, b)) =>
                    val newLin = lin.patch(idx, List((s"$name.1", a), (s"$name.2", b)), 1)
                    Right(state.copy(goals = goal.copy(linearContext = newLin) :: state.goals.tail))
                  case _ => Left(s"destruct: Linear hypothesis '$hypName' cannot be destructed")
                }
              case None => Left(s"destruct: Hypothesis '$hypName' not found")
            }
        }
      case None => Left("No current goal")
    }
  }

  /** 自然数に関する帰納法。
    */
  def induction(
      state: ProofState,
      varName: Option[String] = None
  ): TacticResult = {
    state.currentGoal match {
      case Some(Goal(ctx, lin, App(Sym(Forall), List(Var(v), body)))) =>
        val targetVar = varName.getOrElse(v)
        val baseGoal = Goal(ctx, lin, Prover.substVar(body, v, Sym(Zero)))
        val n = targetVar
        val pn = Prover.substVar(body, v, Var(n))
        val psn = Prover.substVar(body, v, App(Sym(Succ), List(Var(n))))
        val stepGoal = Goal(
          ctx,
          lin,
          App(Sym(Forall), List(Var(n), App(Sym(Implies), List(pn, psn))))
        )
        Right(state.copy(goals = baseGoal :: stepGoal :: state.goals.tail))
      case _ => Left("induction: Goal is not a universal quantification")
    }
  }

  /** 等式の反射律。
    */
  def reflexivity(state: ProofState): TacticResult = {
    state.currentGoal match {
      case Some(Goal(_, _, Sym(True))) =>
        // すでに Simpl などで True に簡約されている場合
        Right(state.copy(goals = state.goals.tail))
      case Some(Goal(_, _, App(Sym(Eq | Path), args))) if args.length >= 2 =>
        val l = args match { case List(_, x, _) if args.length == 3 => x; case List(x, _) => x; case _ => null }
        val r = args.last
        if (l != null && Rewriter.normalize(l) == Rewriter.normalize(r))
          Right(state.copy(goals = state.goals.tail))
        else
          Left(
            s"reflexivity: ${Rewriter.normalize(l)} and ${Rewriter.normalize(r)} are not equal"
          )
      case _ => Left("reflexivity: Goal is not an equality or True")
    }
  }

  /** 仮定にある等式を使って結論を書き換える（高度なインスタンス化対応版）。
    */
  def rewrite(state: ProofState, hypName: String): TacticResult = {
    state.currentGoal match {
      case Some(goal @ Goal(ctx, lin, target)) =>
        val normTarget = Rewriter.normalize(target)
        if (normTarget == Sym(True)) {
          // 目標がすでに解決されている場合
          return Right(state.copy(goals = state.goals.tail))
        }

        (ctx ++ lin).find(_._1 == hypName) match {
          case Some((_, hypExpr)) =>
            // 1. 仮定をインスタンス化
            var metaCounter = 0
            def instantiate(expr: Expr): Expr = expr match {
              case App(Sym(Forall), List(Var(v), body)) =>
                metaCounter += 1
                instantiate(
                  Prover.substVar(body, v, Meta(MetaId(List(999, metaCounter))))
                )
              case _ => expr
            }

            val instHyp = Rewriter.normalize(instantiate(hypExpr))
            instHyp match {
              case App(Sym(Eq | Path), args) if args.length >= 2 =>
                val l = args match { case List(_, x, _) if args.length == 3 => x; case List(x, _) => x; case _ => null }
                val r = args.last

                // 2. 単一化を伴う部分式書き換え
                def findAndReplace(expr: Expr): Option[(Expr, Subst)] = {
                  unify(expr, l, emptySubst).headOption
                    .map(s => (applySubst(r, s), s))
                    .orElse {
                      expr match {
                        case App(f, as) =>
                          as.indices
                            .to(LazyList)
                            .flatMap { idx =>
                              findAndReplace(as(idx)).map {
                                case (newArg, s) =>
                                  val nextF = applySubst(f, s)
                                  val nextArgs = as
                                    .patch(idx, List(newArg), 1)
                                    .map(applySubst(_, s))
                                  (App(nextF, nextArgs), s)
                              }
                            }
                            .headOption
                        case _ => None
                      }
                    }
                }

                findAndReplace(normTarget) match {
                  case Some((rewritten, subst)) =>
                    val finalTarget =
                      Rewriter.normalize(applySubst(rewritten, subst))
                    if (finalTarget == Sym(True))
                      Right(state.copy(goals = state.goals.tail))
                    else
                      Right(
                        state.copy(goals =
                          goal.copy(target = finalTarget) :: state.goals.tail
                        )
                      )
                  case None =>
                    Left(
                      s"rewrite: Hypothesis '$hypName' ($instHyp) not applicable to goal: $normTarget"
                    )
                }
              case Sym(True) => 
                // 仮定自体が True なら、書き換えには使えないがエラーにもしない
                Right(state)
              case _ =>
                Left(s"rewrite: Hypothesis '$hypName' ($instHyp) is not an equality")
            }
          case None => Left(s"rewrite: Hypothesis '$hypName' not found")
        }
      case None => Left("No current goal")
    }
  }

  def exact(state: ProofState, hypName: String): TacticResult = {
    state.currentGoal match {
      case Some(Goal(ctx, lin, target)) =>
        (ctx ++ lin).find(_._1 == hypName) match {
          case Some((_, hypExpr)) =>
            unify(hypExpr, target, emptySubst).headOption match {
              case Some(_) => Right(state.copy(goals = state.goals.tail))
              case None    =>
                Left(s"exact: Hypothesis '$hypName' does not match goal")
            }
          case None => Left(s"exact: Hypothesis '$hypName' not found")
        }
      case None => Left("No current goal")
    }
  }

  def assumption(state: ProofState): TacticResult = {
    state.currentGoal match {
      case Some(Goal(ctx, lin, target)) =>
        if (
          (ctx ++ lin).exists { case (_, hyp) =>
            unify(hyp, target, emptySubst).nonEmpty
          }
        )
          Right(state.copy(goals = state.goals.tail))
        else Left("assumption: No matching hypothesis found")
      case None => Left("No current goal")
    }
  }

  def applyHyp(state: ProofState, hypName: String): TacticResult = {
    state.currentGoal match {
      case Some(Goal(ctx, lin, target)) =>
        ctx.find(_._1 == hypName) match {
          case Some((_, hypExpr)) =>
            hypExpr match {
              case App(Sym(op), List(a, b))
                  if op == Implies || op == ImpliesAlt1 || op == ImpliesAlt2 =>
                unify(b, target, emptySubst).headOption match {
                  case Some(_) =>
                    Right(state.copy(goals = Goal(ctx, lin, a) :: state.goals.tail))
                  case None =>
                    Left(s"apply: Conclusion of '$hypName' does not match goal")
                }
              case _ =>
                Left(s"apply: Hypothesis '$hypName' is not an implication")
            }
          case None => Left(s"apply: Hypothesis '$hypName' not found")
        }
      case None => Left("No current goal")
    }
  }

  /** 全自動探索による解決。
    */
  def auto(state: ProofState, prover: Prover): TacticResult = {
    state.currentGoal match {
      case Some(goal) =>
        prover.prove(goal.target, initialGoal = Some(goal)) match {
          case Right(result) =>
            println(s"✓ Auto solved: ${result.tree.format(0)}")
            Right(
              state.copy(
                goals = state.goals.tail,
                completedProofs = result.tree :: state.completedProofs
              )
            )
          case Left(fail) =>
            Left(s"auto: No proof found. Best attempt depth: ${fail.depth}")
        }
      case None => Left("No current goal")
    }
  }

  /** 目標と仮定の正規化（簡約）。
    */
  def simpl(state: ProofState): TacticResult = {
    state.currentGoal match {
      case Some(goal) =>
        val nextTarget = Rewriter.normalize(goal.target)
        if (nextTarget == Sym(True)) {
          // 簡約の結果 True になったら、そのゴールを解決したとみなす
          Right(state.copy(goals = state.goals.tail))
        } else {
          val nextGoal = Goal(
            goal.context.map((n, e) => (n, Rewriter.normalize(e))),
            goal.linearContext.map((n, e) => (n, Rewriter.normalize(e))),
            nextTarget
          )
          Right(state.copy(goals = nextGoal :: state.goals.tail))
        }
      case None => Left("No current goal")
    }
  }
}

File: Zig/src/prover/tactics.zig
// ==========================================
// tactics.zig
// タクティクスの実装 (Tactics.scala移植)
// intro, split, destruct, auto等
// ==========================================

const std = @import("std");
const Allocator = std.mem.Allocator;
const expr_mod = @import("expr.zig");
const unifier_mod = @import("unifier.zig");
const search_mod = @import("search.zig");
const syms = @import("symbols.zig");
const Expr = expr_mod.Expr;
const Goal = expr_mod.Goal;
const ContextEntry = expr_mod.ContextEntry;
const ProofTree = expr_mod.ProofTree;
const ProofState = expr_mod.ProofState;
const LogicState = expr_mod.LogicState;

pub const TacticError = error{ InvalidGoal, HypothesisNotFound, NoCurrentGoal, OutOfMemory, Timeout };
pub const TacticResult = union(enum) {
    ok: ProofState,
    err: []const u8,
};

/// intro: →, ⊸, ∀ の導入
pub fn intro(state: ProofState, name: ?[]const u8, arena: Allocator) TacticResult {
    const goal = state.currentGoal() orelse return .{ .err = "No current goal" };
    const target = goal.target;

    if (target.* != .app or target.app.head.* != .sym) return .{ .err = "intro: Goal is not an implication or universal quantification" };
    const hn = target.app.head.sym;
    const args = target.app.args;

    // →導入
    if ((std.mem.eql(u8, hn, syms.Implies) or std.mem.eql(u8, hn, "⊃") or std.mem.eql(u8, hn, "⇒")) and args.len == 2) {
        const hyp_name = name orelse blk: {
            var buf: [32]u8 = undefined;
            const n = std.fmt.bufPrint(&buf, "h{d}", .{goal.context.len}) catch return .{ .err = "intro: name gen failed" };
            break :blk arena.dupe(u8, n) catch return .{ .err = "OOM" };
        };
        var new_ctx: std.ArrayList(ContextEntry) = .{};
        new_ctx.append(arena, .{ .name = hyp_name, .expr = args[0] }) catch return .{ .err = "OOM" };
        for (goal.context) |e| new_ctx.append(arena, e) catch return .{ .err = "OOM" };

        const new_goal = Goal{ .context = new_ctx.items, .linear_context = goal.linear_context, .target = args[1] };
        var new_goals: std.ArrayList(Goal) = .{};
        new_goals.append(arena, new_goal) catch return .{ .err = "OOM" };
        for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
        return .{ .ok = .{
            .goals = new_goals.items,
            .completed_proofs = state.completed_proofs,
            .original_goal = state.original_goal,
            .rules = state.rules,
        } };
    }

    // ⊸導入
    if (std.mem.eql(u8, hn, syms.LImplies) and args.len == 2) {
        const hyp_name = name orelse blk: {
            var buf: [32]u8 = undefined;
            const n = std.fmt.bufPrint(&buf, "lin{d}", .{goal.linear_context.len}) catch return .{ .err = "intro: name gen failed" };
            break :blk arena.dupe(u8, n) catch return .{ .err = "OOM" };
        };
        var new_lin: std.ArrayList(ContextEntry) = .{};
        new_lin.append(arena, .{ .name = hyp_name, .expr = args[0] }) catch return .{ .err = "OOM" };
        for (goal.linear_context) |e| new_lin.append(arena, e) catch return .{ .err = "OOM" };

        const new_goal = Goal{ .context = goal.context, .linear_context = new_lin.items, .target = args[1] };
        var new_goals: std.ArrayList(Goal) = .{};
        new_goals.append(arena, new_goal) catch return .{ .err = "OOM" };
        for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
        return .{ .ok = .{
            .goals = new_goals.items,
            .completed_proofs = state.completed_proofs,
            .original_goal = state.original_goal,
            .rules = state.rules,
        } };
    }

    // ∀導入
    if (std.mem.eql(u8, hn, syms.Forall) and args.len >= 2 and args[0].* == .var_) {
        const v_name = args[0].var_;
        const body = if (args.len == 3) args[2] else args[1];
        const hyp_name = name orelse blk: {
            var buf: [32]u8 = undefined;
            const n = std.fmt.bufPrint(&buf, "{s}_t{d}", .{ v_name, state.goals.len }) catch return .{ .err = "intro: name gen failed" };
            break :blk arena.dupe(u8, n) catch return .{ .err = "OOM" };
        };
        const fresh_var = expr_mod.var_(arena, hyp_name) catch return .{ .err = "OOM" };
        const instantiated = unifier_mod.substVar(body, v_name, fresh_var, arena) catch return .{ .err = "OOM" };
        const new_goal = Goal{ .context = goal.context, .linear_context = goal.linear_context, .target = instantiated };
        var new_goals: std.ArrayList(Goal) = .{};
        new_goals.append(arena, new_goal) catch return .{ .err = "OOM" };
        for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
        return .{ .ok = .{
            .goals = new_goals.items,
            .completed_proofs = state.completed_proofs,
            .original_goal = state.original_goal,
            .rules = state.rules,
        } };
    }

    return .{ .err = "intro: Goal is not an implication or universal quantification" };
}

/// split: ∧の分解
pub fn split(state: ProofState, arena: Allocator) TacticResult {
    const goal = state.currentGoal() orelse return .{ .err = "No current goal" };
    const target = goal.target;

    if (target.* != .app or target.app.head.* != .sym) return .{ .err = "split: Goal is not a conjunction" };
    const hn = target.app.head.sym;
    const args = target.app.args;

    if ((std.mem.eql(u8, hn, syms.And) or std.mem.eql(u8, hn, syms.Product)) and args.len == 2) {
        const g1 = Goal{ .context = goal.context, .linear_context = goal.linear_context, .target = args[0] };
        const g2 = Goal{ .context = goal.context, .linear_context = goal.linear_context, .target = args[1] };
        var new_goals: std.ArrayList(Goal) = .{};
        new_goals.append(arena, g1) catch return .{ .err = "OOM" };
        new_goals.append(arena, g2) catch return .{ .err = "OOM" };
        for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
        return .{ .ok = .{
            .goals = new_goals.items,
            .completed_proofs = state.completed_proofs,
            .original_goal = state.original_goal,
            .rules = state.rules,
        } };
    }

    return .{ .err = "split: Goal is not a conjunction" };
}

/// auto: 自動証明を試行
pub fn auto(state: ProofState, plugins: []const search_mod.Plugin, arena: Allocator) TacticResult {
    const goal = state.currentGoal() orelse return .{ .err = "No current goal" };

    var engine = search_mod.ProverEngine.init(
        expr_mod.ProverConfig{ .rules = state.rules },
        plugins,
        arena,
        arena,
    );
    defer engine.deinit();

    const result = engine.prove(goal.target, 10, 5000) catch return .{ .err = "auto: timeout or error" };

    return switch (result) {
        .success => |s| {
            var new_proofs: std.ArrayList(ProofTree) = .{};
            for (state.completed_proofs) |p| new_proofs.append(arena, p) catch return .{ .err = "OOM" };
            new_proofs.append(arena, s.tree) catch return .{ .err = "OOM" };
            var new_goals: std.ArrayList(Goal) = .{};
            for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
            return .{ .ok = .{
                .goals = new_goals.items,
                .completed_proofs = new_proofs.items,
                .original_goal = state.original_goal,
                .rules = state.rules,
            } };
        },
        .fail => .{ .err = "auto: no proof found" },
    };
}

/// assumption: コンテキストからゴールを直接解決
pub fn assumption(state: ProofState, arena: Allocator) TacticResult {
    const goal = state.currentGoal() orelse return .{ .err = "No current goal" };

    for (goal.context) |entry| {
        if (entry.expr.eql(goal.target)) {
            var new_goals: std.ArrayList(Goal) = .{};
            for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
            return .{ .ok = .{
                .goals = new_goals.items,
                .completed_proofs = state.completed_proofs,
                .original_goal = state.original_goal,
                .rules = state.rules,
            } };
        }
    }

    return .{ .err = "assumption: No matching hypothesis found" };
}

/// reflexivity: 等式 a = a を解決
pub fn reflexivity(state: ProofState, arena: Allocator) TacticResult {
    const goal = state.currentGoal() orelse return .{ .err = "No current goal" };
    const target = goal.target;

    if (target.* == .app and target.app.head.* == .sym and target.app.args.len == 2) {
        const hn = target.app.head.sym;
        if (std.mem.eql(u8, hn, "=") or std.mem.eql(u8, hn, syms.Path)) {
            const l_idx: usize = if (std.mem.eql(u8, hn, syms.Path) and target.app.args.len == 3) 1 else 0;
            const r_idx: usize = if (std.mem.eql(u8, hn, syms.Path) and target.app.args.len == 3) 2 else 1;
            if (target.app.args.len > r_idx and target.app.args[l_idx].eql(target.app.args[r_idx])) {
                var new_goals: std.ArrayList(Goal) = .{};
                for (state.goals[1..]) |g| new_goals.append(arena, g) catch return .{ .err = "OOM" };
                return .{ .ok = .{
                    .goals = new_goals.items,
                    .completed_proofs = state.completed_proofs,
                    .original_goal = state.original_goal,
                    .rules = state.rules,
                } };
            }
        }
    }

    return .{ .err = "reflexivity: Goal is not a reflexive equation" };
}

// テスト
test "intro tactic on implication" {
    var arena_state = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena_state.deinit();
    const arena = arena_state.allocator();

    const implies_sym = try expr_mod.sym(arena, "→");
    const a = try expr_mod.sym(arena, "A");
    const b = try expr_mod.sym(arena, "B");
    const target = try expr_mod.app2(arena, implies_sym, a, b);

    const state = ProofState{
        .goals = &[_]Goal{.{ .target = target }},
        .completed_proofs = &.{},
        .original_goal = target,
    };

    const result = intro(state, null, arena);
    switch (result) {
        .ok => |s| {
            try std.testing.expect(s.goals.len == 1);
            try std.testing.expect(s.goals[0].target.eql(b));
            try std.testing.expect(s.goals[0].context.len == 1);
        },
        .err => |msg| {
            std.debug.print("Unexpected error: {s}\n", .{msg});
            try std.testing.expect(false);
        },
    }
}
